File: Src/LoLMatchAccepterNet/LCU/MatchAcceptService.cs
namespace LoLMatchAccepterNet.LCU
{
    public class MatchAcceptService
    {
        private readonly Game _game;

        public MatchAcceptService(Game game)
        {
            _game = game;
        }

        public async Task RunAutoAcceptLoopAsync(CancellationToken cancellationToken)
        {
            while (!cancellationToken.IsCancellationRequested)
            {
                try
                {
                    bool isMatchAccepted = await _game.WaitForQueue();

                    if (isMatchAccepted)
                    {
                        string currentPhase = await _game.WaitUntilPhaseEnds(Game.ReadyCheck);

                        if (currentPhase == Game.ChampSelect)
                        {
                            Console.WriteLine("Currently in champ select...");
                            await _game.WaitUntilPhaseEnds(Game.ChampSelect);
                        }
                    }

                    bool inGame = await _game.IsActive();

                    if (inGame)
                    {
                        Console.WriteLine("Active game detected. Waiting until game ends...");
                        await _game.WaitUntilGameEnds();
                        Console.WriteLine("Game ended. Navigating to lobby...");
                        await _game.NavigateToLobby();
                        Console.WriteLine("Resuming auto-accept...");
                    }
                }
                catch (OperationCanceledException)
                {
                    throw;
                }
                catch (Exception)
                {
                    Console.WriteLine($"Error communicating with League client.");
                    await Task.Delay(6000, cancellationToken);
                    throw;
                }

                await Task.Delay(500, cancellationToken);
            }

            cancellationToken.ThrowIfCancellationRequested();
        }
    }
}

File: Src/LoLMatchAccepterCpp/src/LCU/MatchAcceptService.cpp
#include "MatchAcceptService.h"
#include <iostream>
#include <thread>
#include <chrono>

MatchAcceptService::MatchAcceptService(std::shared_ptr<Game> game) : _game(game) {
}

void MatchAcceptService::RunAutoAcceptLoop(std::atomic<bool>& cancelToken) {
    while (!cancelToken.load()) {
        try {
            bool isMatchAccepted = _game->WaitForQueue();

            if (isMatchAccepted) {
                std::string currentPhase = _game->WaitUntilPhaseEnds({Game::ReadyCheck});

                if (currentPhase == Game::ChampSelect) {
                    std::cout << "Currently in champ select..." << std::endl;
                    _game->WaitUntilPhaseEnds({Game::ChampSelect});
                }
            }

            bool inGame = _game->IsActive();

            if (inGame) {
                std::cout << "Active game detected. Waiting until game ends..." << std::endl;
                _game->WaitUntilGameEnds();
                std::cout << "Game ended. Navigating to lobby..." << std::endl;
                _game->NavigateToLobby();
                std::cout << "Resuming auto-accept..." << std::endl;
            }
        }
        catch (const std::exception&) {
            std::cout << "Error communicating with League client." << std::endl;
            std::this_thread::sleep_for(std::chrono::milliseconds(6000));
            throw;
        }

        std::this_thread::sleep_for(std::chrono::milliseconds(500));
    }
}
